Sources/GRPCHTTP2Core/Client/Connection/ClientConnectionHandler.swift

```diff
@@ -0,0 +1,421 @@
+    self.keepAliveTimeoutTimer = Timer(delay: keepAliveTimeout ?? .seconds(20))
+    self.keepAlivePingData = HTTP2PingData(withInteger: .random(in: .min ... .max))
+    self.state = StateMachine(allowKeepAliveWithoutCalls: keepAliveWithoutCalls)
+
+    self.flushPending = false
+    self.inReadLoop = false
+  }
+
+  func handlerAdded(context: ChannelHandlerContext) {
+    assert(context.eventLoop === self.eventLoop)
+  }
+
+  func channelActive(context: ChannelHandlerContext) {
+    self.keepAliveTimer?.schedule(on: context.eventLoop) {
+      self.keepAliveTimerFired(context: context)
+    }
+
+    self.maxIdleTimer?.schedule(on: context.eventLoop) {
+      self.maxIdleTimerFired(context: context)
+    }
+  }
+
+  func channelInactive(context: ChannelHandlerContext) {
+    self.state.closed()
+    self.keepAliveTimer?.cancel()
+    self.keepAliveTimeoutTimer.cancel()
+  }
+
+  func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
+    switch event {
+    case let event as NIOHTTP2StreamCreatedEvent:
+      // Stream created, so the connection isn't idle.
+      self.maxIdleTimer?.cancel()
+      self.state.streamOpened(event.streamID)
+
+    case let event as StreamClosedEvent:
+      switch self.state.streamClosed(event.streamID) {
+      case .startIdleTimer(let cancelKeepAlive):
+        // All streams are closed, restart the idle timer, and stop the keep-alive timer (it may
+        // not stop if keep-alive is allowed when there are no active calls).
+        self.maxIdleTimer?.schedule(on: context.eventLoop) {
+          self.maxIdleTimerFired(context: context)
+        }
+
+        if cancelKeepAlive {
+          self.keepAliveTimer?.cancel()
+        }
+
+      case .close:
+        // Connection was closing but waiting for all streams to close. They must all be closed
+        // now so close the connection.
+        context.close(promise: nil)
+
+      case .none:
+        ()
+      }
+
+    default:
+      ()
+    }
+
+    context.fireUserInboundEventTriggered(event)
+  }
+
+  func channelRead(context: ChannelHandlerContext, data: NIOAny) {
+    let frame = self.unwrapInboundIn(data)
+    self.inReadLoop = true
+
+    switch frame.payload {
+    case .goAway(_, let errorCode, let data):
+      // Receiving a GOAWAY frame means we need to stop creating streams immediately and start
+      // closing the connection.
+      switch self.state.beginGracefulShutdown() {
+      case .sendGoAway(let close):
+        // gRPC servers may indicate why the GOAWAY was sent in the opaque data.
+        let message = data.map { String(buffer: $0) } ?? ""
+        context.fireChannelRead(self.wrapInboundOut(.closing(.goAway(errorCode, message))))
+
+        // Clients should send GOAWAYs when closing a connection.
+        self.writeAndFlushGoAway(context: context, errorCode: .noError)
+        if close {
+          context.close(promise: nil)
+        }
```

Comment:
No, GOAWAY means the connection will be closed in the future, it doesn't have to be closed immediately. 

In this case the client is reacting to a GOAWAY from the server, if there are open streams then the client lets those continue to run otherwise the client closes immediately as there's nothing left to do.